x/avs/keeper/avs.go

```diff
@@ -37,25 +54,74 @@ func (k *Keeper) GetAVSMinimumSelfDelegation(ctx sdk.Context, avsAddr string) (s
 	}
 
-	return sdkmath.LegacyNewDec(avsInfo.Info.MinSelfDelegation.Int64()), nil
+	return sdkmath.LegacyNewDec(int64(avsInfo.Info.MinSelfDelegation)), nil
 }
 
 // GetEpochEndAVSs returns the AVS list where the current block marks the end of their epoch.
-func (k *Keeper) GetEpochEndAVSs(ctx sdk.Context) ([]string, error) {
-	var avsList []types.AVSInfo
-	k.IterateAVSInfo(ctx, func(_ int64, epochEndAVSInfo types.AVSInfo) (stop bool) {
-		avsList = append(avsList, epochEndAVSInfo)
+func (k *Keeper) GetEpochEndAVSs(ctx sdk.Context, epochIdentifier string, epochNumber int64) ([]string, error) {
+	var avsList []string
+	k.IterateAVSInfo(ctx, func(_ int64, avsInfo types.AVSInfo) (stop bool) {
+		if epochIdentifier == avsInfo.EpochIdentifier && epochNumber > int64(avsInfo.StartingEpoch) {
+			avsList = append(avsList, avsInfo.AvsAddress)
+		}
+		return false
+	})
+
+	return avsList, nil
+}
+
+// GetTaskChallengeEpochEndAVSs returns the AVS list where the current block marks the end of their epoch.
+func (k *Keeper) GetTaskChallengeEpochEndAVSs(ctx sdk.Context, epochIdentifier string, epochNumber int64) ([]string, error) {
+	var avsList []string
+	k.IterateAVSInfo(ctx, func(_ int64, avsInfo types.AVSInfo) (stop bool) {
+		if epochIdentifier == avsInfo.EpochIdentifier && epochNumber > int64(avsInfo.StartingEpoch) {
+			avsList = append(avsList, avsInfo.AvsAddress)
+		}
 		return false
 	})
+	return avsList, nil
+}
 
-	if len(avsList) == 0 {
-		return []string{}, nil
+func (k *Keeper) GetAVSAddrByChainID(ctx sdk.Context, chainID string) (string, error) {
+	chainID = ChainIDWithoutRevision(chainID)
+	if len(chainID) == 0 {
+		return "", errorsmod.Wrap(types.ErrNotNull, "RegisterAVSWithChainID: chainID is null")
+	}
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixAVSInfoByChainID)
+	if !store.Has([]byte(chainID)) {
+		return "", errorsmod.Wrap(types.ErrNoKeyInTheStore, fmt.Sprintf("GetAVSAddrByChainID: key is %s", chainID))
 	}
+	avsAddr := store.Get([]byte(chainID))
+
+	return string(avsAddr), nil
+}
 
-	avsAddrList := make([]string, len(avsList))
-	for i := range avsList {
-		avsAddrList[i] = avsList[i].AvsAddress
+// RegisterAVSWithChainID creates an avs address given the chainID
+func (k Keeper) RegisterAVSWithChainID(ctx sdk.Context, chainID string) (err error) {
+	chainID = ChainIDWithoutRevision(chainID)
+	if len(chainID) == 0 {
+		return errorsmod.Wrap(err, "RegisterAVSWithChainID: chainID is null")
 	}
+	avsAddr := common.BytesToAddress(crypto.Keccak256([]byte(chainID))).String()
+
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixAVSInfoByChainID)
 
-	return avsAddrList, nil
+	err = k.evmKeeper.SetAccount(ctx, common.HexToAddress(avsAddr), statedb.Account{
+		Balance:  big.NewInt(0),
+		CodeHash: crypto.Keccak256Hash([]byte(types.ChainID)).Bytes(),
+		Nonce:    1,
+	})
+	if err != nil {
+		return err
+	}
+	store.Set([]byte(chainID), []byte(avsAddr))
```

Comment:
Can you store the reverse lookup also so that the `chainID` can be pulled from the `avsAddr` ? And create the associated function.